### PR TITLE
Initial Commit of Bootstrap v4 Snippets. Part 1

### PR DESCRIPTION
**Pull request by mwweb**

### CODE REVIEW

**Comment by mwweb** [comment]
@Llewellynvdm This is the first pull request of the first 50. If these look good, then I will start on the others.

File: Bootstrap v4 - (Breadcrumbs) Breadcrumb.json [hunk 1]
```diff
@@ -0,0 +1,16 @@
+{
+    "name": "Breadcrumb",
+    "heading": "breadcrumb",
+    "description": "Indicate the current page's location within a navigational hierarchy that automatically adds separators via CSS.",
+    "usage": "Separators are automatically added in CSS through ::before and content.",
+    "snippet": "<ol class=\"\"breadcrumb\"\">\r\n  <li class=\"\"breadcrumb-item active\"\">Home<\/li>\r\n<\/ol>\r\n<ol class=\"\"breadcrumb\"\">\r\n  <li class=\"\"breadcrumb-item\"\"><a href=\"\"#!\"\">Home<\/a><\/li>\r\n  <li class=\"\"breadcrumb-item active\"\">Library<\/li>\r\n<\/ol>\r\n<ol class=\"\"breadcrumb\"\">\r\n  <li class=\"\"breadcrumb-item\"\"><a href=\"\"#!\"\">Home<\/a><\/li>\r\n  <li class=\"\"breadcrumb-item\"\"><a href=\"\"#!\"\">Library<\/a><\/li>\r\n  <li class=\"\"breadcrumb-item active\"\">Data<\/li>\r\n<\/ol>",
```
Comment (Llewellynvdm) [inline Bootstrap v4 - (Breadcrumbs) Breadcrumb.json]: Here is some more of those :(

Comment (mwweb) [reply Bootstrap v4 - (Breadcrumbs) Breadcrumb.json]: Hmm. I just edited that one, again, and it is the only one in the batch that I see that has that issue.

File: Bootstrap v4 - (Button Groups) Button Group (nested).json [hunk 2]
```diff
@@ -0,0 +1,16 @@
+{
+    "name": "Button Group (nested)",
+    "heading": "btn-group (nested)",
+    "description": "Group a series of buttons together on a single line with the button group, and super-power them with JavaScript.",
+    "usage": "Wrap a series of buttons with .btn in .btn-group. Add on optional JavaScript radio and checkbox style behavior with our buttons plugin.",
+    "snippet": "<div class=\"\"btn-group\"\" role=\"\"group\"\" aria-label=\"\"Button group with nested dropdown\"\">\r\n  <button type=\"\"button\"\" class=\"\"btn btn-secondary\"\">1<\/button>\r\n  <button type=\"\"button\"\" class=\"\"btn btn-secondary\"\">2<\/button>\r\n\r\n  <div class=\"\"btn-group\"\" role=\"\"group\"\">\r\n    <button id=\"\"btnGroupDrop1\"\" type=\"\"button\"\" class=\"\"btn btn-secondary dropdown-toggle\"\" data-toggle=\"\"dropdown\"\" aria-haspopup=\"\"true\"\" aria-expanded=\"\"false\"\">\r\n      Dropdown\r\n    <\/button>\r\n    <div class=\"\"dropdown-menu\"\" aria-labelledby=\"\"btnGroupDrop1\"\">\r\n      <a class=\"\"dropdown-item\"\" href=\"\"#!\"\">Dropdown link<\/a>\r\n      <a class=\"\"dropdown-item\"\" href=\"\"#!\"\">Dropdown link<\/a>\r\n    <\/div>\r\n  <\/div>\r\n<\/div>",
```
Comment (Llewellynvdm) [inline Bootstrap v4 - (Button Groups) Button Group (nested).json]: Check

File: Bootstrap v4 - (Button Groups) Button Group - Large.json [hunk 3]
```diff
@@ -0,0 +1,16 @@
+{
+    "name": "Button Group - Large",
+    "heading": "btn-group-lg",
+    "description": "Group a series of buttons together on a single line with the button group, and super-power them with JavaScript.",
+    "usage": "Wrap a series of buttons with .btn in .btn-group. Add on optional JavaScript radio and checkbox style behavior with our buttons plugin.",
+    "snippet": "<div class=\"\"btn-group btn-group-lg\"\" role=\"\"group\"\" aria-label=\"\"Basic example\"\">\r\n  <button type=\"\"button\"\" class=\"\"btn btn-secondary\"\">Left<\/button>\r\n  <button type=\"\"button\"\" class=\"\"btn btn-secondary\"\">Middle<\/button>\r\n  <button type=\"\"button\"\" class=\"\"btn btn-secondary\"\">Right<\/button>\r\n<\/div>",
```
Comment (Llewellynvdm) [inline Bootstrap v4 - (Button Groups) Button Group - Large.json]: Check

File: Bootstrap v4 - (Button Groups) Button Group - Small.json [hunk 4]
```diff
@@ -0,0 +1,16 @@
+{
+    "name": "Button Group - Small",
+    "heading": "btn-group-sm",
+    "description": "Group a series of buttons together on a single line with the button group, and super-power them with JavaScript.",
+    "usage": "Wrap a series of buttons with .btn in .btn-group. Add on optional JavaScript radio and checkbox style behavior with our buttons plugin.",
+    "snippet": "<div class=\"\"btn-group btn-group-sm\"\" role=\"\"group\"\" aria-label=\"\"Basic example\"\">\r\n  <button type=\"\"button\"\" class=\"\"btn btn-secondary\"\">Left<\/button>\r\n  <button type=\"\"button\"\" class=\"\"btn btn-secondary\"\">Middle<\/button>\r\n  <button type=\"\"button\"\" class=\"\"btn btn-secondary\"\">Right<\/button>\r\n<\/div>",
```
Comment (Llewellynvdm) [inline Bootstrap v4 - (Button Groups) Button Group - Small.json]: Check

File: Bootstrap v4 - (Button Groups) Button Group - Vertical.json [hunk 5]
```diff
@@ -0,0 +1,16 @@
+{
+    "name": "Button Group - Vertical",
+    "heading": "btn-group-vertical",
+    "description": "Group a series of buttons together on a single line with the button group, and super-power them with JavaScript.",
+    "usage": "Wrap a series of buttons with .btn in .btn-group. Add on optional JavaScript radio and checkbox style behavior with our buttons plugin.",
+    "snippet": "<div class=\"\"btn-group-vertical\"\" role=\"\"group\"\" aria-label=\"\"Basic example\"\">\r\n  <button type=\"\"button\"\" class=\"\"btn btn-secondary\"\">Left<\/button>\r\n  <button type=\"\"button\"\" class=\"\"btn btn-secondary\"\">Middle<\/button>\r\n  <button type=\"\"button\"\" class=\"\"btn btn-secondary\"\">Right<\/button>\r\n<\/div>",
```
Comment (Llewellynvdm) [inline Bootstrap v4 - (Button Groups) Button Group - Vertical.json]: Check

File: Bootstrap v4 - (Button Groups) Button Group.json [hunk 6]
```diff
@@ -0,0 +1,16 @@
+{
+    "name": "Button Group",
+    "heading": "btn-group\u00a0",
+    "description": "Group a series of buttons together on a single line with the button group, and super-power them with JavaScript.",
+    "usage": "Wrap a series of buttons with .btn in .btn-group. Add on optional JavaScript radio and checkbox style behavior with our buttons plugin.",
+    "snippet": "<div class=\"\"btn-group\"\" role=\"\"group\"\" aria-label=\"\"Basic example\"\">\r\n  <button type=\"\"button\"\" class=\"\"btn btn-secondary\"\">Left<\/button>\r\n  <button type=\"\"button\"\" class=\"\"btn btn-secondary\"\">Middle<\/button>\r\n  <button type=\"\"button\"\" class=\"\"btn btn-secondary\"\">Right<\/button>\r\n<\/div>",
```
Comment (Llewellynvdm) [inline Bootstrap v4 - (Button Groups) Button Group.json]: Check

File: Bootstrap v4 - (Button Modifiers) Button - As Checkbox.json [hunk 7]
```diff
@@ -0,0 +1,16 @@
+{
+    "name": "Button - As Checkbox",
+    "heading": "checkbox as button",
+    "description": "Use Bootstrap's custom button styles for actions in forms, dialogs, and more with support for multiple sizes, states, and more.",
+    "usage": "Fancy larger or smaller buttons? Add .btn-lg or .btn-sm for additional sizes.",
+    "snippet": "<div class=\"\"btn-group\"\" data-toggle=\"\"buttons\"\">\r\n  <label class=\"\"btn btn-primary active\"\">\r\n    <input type=\"\"checkbox\"\" checked=\"\"\"\"> Checkbox 1 (pre-checked)\r\n  <\/label>\r\n  <label class=\"\"btn btn-primary\"\">\r\n    <input type=\"\"checkbox\"\"> Checkbox 2\r\n  <\/label>\r\n<\/div>",
```
Comment (Llewellynvdm) [inline Bootstrap v4 - (Button Modifiers) Button - As Checkbox.json]: Check

File: Bootstrap v4 - (Button Modifiers) Button - As Radio.json [hunk 8]
```diff
@@ -0,0 +1,16 @@
+{
+    "name": "Button - As Radio",
+    "heading": "radio as button",
+    "description": "Use Bootstrap's custom button styles for actions in forms, dialogs, and more with support for multiple sizes, states, and more.",
+    "usage": "Fancy larger or smaller buttons? Add .btn-lg or .btn-sm for additional sizes.",
+    "snippet": "<div class=\"\"btn-group\"\" data-toggle=\"\"buttons\"\">\r\n  <label class=\"\"btn btn-primary active\"\">\r\n    <input type=\"\"radio\"\" name=\"\"options\"\" id=\"\"option1\"\" checked=\"\"\"\"> Radio 1 (preselected)\r\n  <\/label>\r\n  <label class=\"\"btn btn-primary\"\">\r\n    <input type=\"\"radio\"\" name=\"\"options\"\" id=\"\"option2\"\"> Radio 2\r\n  <\/label>\r\n<\/div>",
```
Comment (Llewellynvdm) [inline Bootstrap v4 - (Button Modifiers) Button - As Radio.json]: Check

File: Bootstrap v4 - (Breadcrumbs) Breadcrumb.json (hunk 1 above)
Comment (Llewellynvdm) [reply Bootstrap v4 - (Breadcrumbs) Breadcrumb.json]: I have seen more, I marked a few...

**Comment by mwweb** [comment]
OK. All of the snippets in this batch have been checked, and edited as necessary. I'm going through the others looking for them stray double quotes.

File: Bootstrap v4 - (Button Groups) Button - Toolbar.json [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+    "usage": "Combine sets of button groups into button toolbars for more complex components. Use utility classes as needed to space out groups, buttons, and more.",
+    "snippet": "<div class=\"\"btn-toolbar\"\" role=\"\"toolbar\"\" aria-label=\"\"Toolbar with button groups\"\">\r\n  <div class=\"\"btn-group\"\" role=\"\"group\"\" aria-label=\"\"First group\"\">\r\n    <button type=\"\"button\"\" class=\"\"btn btn-secondary\"\">1<\/button>\r\n    <button type=\"\"button\"\" class=\"\"btn btn-secondary\"\">2<\/button>\r\n    <button type=\"\"button\"\" class=\"\"btn btn-secondary\"\">3<\/button>\r\n  <\/div>\r\n  <div class=\"\"btn-group\"\" role=\"\"group\"\" aria-label=\"\"Second group\"\">\r\n    <button type=\"\"button\"\" class=\"\"btn btn-secondary\"\">5<\/button>\r\n    <button type=\"\"button\"\" class=\"\"btn btn-secondary\"\">6<\/button>\r\n  <\/div>\r\n  <div class=\"\"btn-group\"\" role=\"\"group\"\" aria-label=\"\"Third group\"\">\r\n    <button type=\"\"button\"\" class=\"\"btn btn-secondary\"\">8<\/button>\r\n  <\/div>\r\n<\/div>",
```
Comment (Llewellynvdm) [inline Bootstrap v4 - (Button Groups) Button - Toolbar.json:6]: heads-up

File: Bootstrap v4 - (Cards) Card - Body.json [hunk 10]
```diff
@@ -0,0 +1,16 @@
+{
+    "name": "Card - Body",
+    "heading": "card-body",
+    "description": "A card is a flexible and extensible content container. It includes options for headers and footers, a wide variety of content, contextual background colors, and powerful display options.\r\n\r\nIf you\u2019re familiar with Bootstrap 3, cards replace our old panels, wells, and thumbnails. Similar functionality to those components is available as modifier classes for cards.",
+    "usage": "Cards are built with as little markup and styles as possible, but still manage to deliver a ton of control and customization. Built with flexbox, they offer easy alignment and mix well with other Bootstrap components.\r\n\r\nCards have no fixed width to start, so they\u2019ll naturally fill the full width of its parent element. This is easily customized with our various sizing options.\r\n\r\n",
+    "snippet": "<div class=\"\"card\"\">\r\n  <div class=\"\"card-body\"\">\r\n    This is some text within a card block.\r\n  <\/div>\r\n<\/div>",
```
Comment (Llewellynvdm) [inline Bootstrap v4 - (Cards) Card - Body.json:6]: heads-up

File: Bootstrap v4 - (Cards) Card.json [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+    "usage": "Cards are built with as little markup and styles as possible, but still manage to deliver a ton of control and customization. Built with flexbox, they offer easy alignment and mix well with other Bootstrap components.\r\n\r\nCards have no fixed width to start, so they\u2019ll naturally fill the full width of its parent element. This is easily customized with our various sizing options.\r\n\r\n",
+    "snippet": "<div class=\"\"card\"\">\r\n  <img class=\"\"card-img-top\"\" src=\"\"\/images\/pathToYourImage.png\"\" alt=\"\"Card image cap\"\">\r\n  <div class=\"\"card-body\"\">\r\n    <h4 class=\"\"card-title\"\">Card title<\/h4>\r\n    <p class=\"\"card-text\"\">\r\n      Some quick example text to build on the card title\r\n      and make up the bulk of the card's content.\r\n    <\/p>\r\n    <a href=\"\"#!\"\" class=\"\"btn btn-primary\"\">Go somewhere<\/a>\r\n  <\/div>\r\n<\/div>",
```
Comment (Llewellynvdm) [inline Bootstrap v4 - (Cards) Card.json:6]: heads-up

File: Bootstrap v4 - (Button Modifiers) Button - As Checkbox.json [hunk 12]
```diff
@@ -0,0 +1,16 @@
+{
+    "name": "Button - As Checkbox",
+    "heading": "checkbox as button",
+    "description": "Use Bootstrap's custom button styles for actions in forms, dialogs, and more with support for multiple sizes, states, and more.",
+    "usage": "Fancy larger or smaller buttons? Add .btn-lg or .btn-sm for additional sizes.",
+    "snippet": "<div class=\"btn-group\" data-toggle=\"buttons\">\r\n  <label class=\"btn btn-primary active\">\r\n    <input type=\"checkbox\" checked=\"\"> Checkbox 1 (pre-checked)\r\n  <\/label>\r\n  <label class=\"btn btn-primary\">\r\n    <input type=\"checkbox\"> Checkbox 2\r\n  <\/label>\r\n<\/div>",
```
Comment (Llewellynvdm) [inline Bootstrap v4 - (Button Modifiers) Button - As Checkbox.json:6]: heads-up

File: Bootstrap v4 - (Button Modifiers) Button - As Radio.json [hunk 13]
```diff
@@ -0,0 +1,16 @@
+{
+    "name": "Button - As Radio",
+    "heading": "radio as button",
+    "description": "Use Bootstrap's custom button styles for actions in forms, dialogs, and more with support for multiple sizes, states, and more.",
+    "usage": "Fancy larger or smaller buttons? Add .btn-lg or .btn-sm for additional sizes.",
+    "snippet": "<div class=\"btn-group\" data-toggle=\"buttons\">\r\n  <label class=\"btn btn-primary active\">\r\n    <input type=\"radio\" name=\"options\" id=\"option1\" checked=\"\"> Radio 1 (preselected)\r\n  <\/label>\r\n  <label class=\"btn btn-primary\">\r\n    <input type=\"radio\" name=\"options\" id=\"option2\"> Radio 2\r\n  <\/label>\r\n<\/div>",
```
Comment (Llewellynvdm) [inline Bootstrap v4 - (Button Modifiers) Button - As Radio.json:6]: heads-up

Comment (Llewellynvdm) [reply Bootstrap v4 - (Button Modifiers) Button - As Radio.json:6]: oops false alarm here

File: Bootstrap v4 - (Button Modifiers) Button - As Checkbox.json (hunk 12 above)
Comment (Llewellynvdm) [reply Bootstrap v4 - (Button Modifiers) Button - As Checkbox.json:6]: oops false alarm here

**Comment by Llewellynvdm** [comment]
I can give a perl script to do a bulk fix on the rest...

**Comment by Llewellynvdm** [comment]
Something like
```
perl -pi -w -e 's/\\"\\"//g' *.json
```
Not tested... so make a backup first

**Comment by Llewellynvdm** [comment]
hey that was wrong... it will completely remove it
```
perl -pi -w -e 's/\\"\\"/\\"/g' *.json
```
Should work...

**Comment by Llewellynvdm** [comment]
Okay so I quickly tested that last command and it works great, you will just have to also run this to fix the empty URLs
```
perl -pi -w -e 's/href=\\"/href=\\"\\"/g' *.json
```
after running 
```
perl -pi -w -e 's/\\"\\"/\\"/g' *.json
```

**Comment by Llewellynvdm** [comment]
[Joomla-Component-Builder-Snippets-fixed.zip](https://github.com/vdm-io/Joomla-Component-Builder-Snippets/files/1493569/Joomla-Component-Builder-Snippets-fixed.zip)

Here is all the snippets fixed. You can just unzip it over the old files.

**Comment by Llewellynvdm** [comment]
wow just found another problem... that second command needs a third. One moment.. here is the correct one.

[Joomla-Component-Builder-Snippets-fix.zip](https://github.com/vdm-io/Joomla-Component-Builder-Snippets/files/1493574/Joomla-Component-Builder-Snippets-fix.zip)

This was pretty nasty, but I think it should now be resolved.

**Comment by Llewellynvdm** [comment]
Tell me are you able to do Skype? I would like to speak to you if that is possible.

**Comment by Llewellynvdm** [comment]
I have an idea how we can reverse this and basically start over... doing all 300+ snippets at once and just get this behind us.

**Comment by mwweb** [comment]
I'll close this, revert mine, then redo it with all of them.

**Pull request closed without merging**